File: pto/core/base/tracer.py

class Tracer:
   """
   A tracing mechanism for reproducible stochastic operations.
   
   The Tracer enables recording and replaying sequences of stochastic operations. When active, 
   it traces all sampling operations and can replay them exactly using the recorded trace.
   
   Attributes:
       active (bool): Whether tracing is currently active
       input_trace (dict): Map of names to distributions from previous trace
       output_trace (dict): Map of names to distributions being recorded
   """
   def __init__(self):
       """Initialize an inactive tracer with empty traces."""
       self.active = False
       self.input_trace = {}
       self.output_trace = {}
   
   def sample(self, name, dist):
       """
       Sample from a distribution with optional tracing.
       
       When tracing is active, this either:
       - Uses a value from the input trace if the distribution matches
       - Repairs the distribution using the traced value if different
       - Samples a new value if no trace exists
       The distribution is recorded in the output trace.
       
       Args:
           name (str): Unique identifier for this sampling operation
           dist (Dist): Distribution to sample from
           
       Returns:
           The sampled value
           
       Raises:
           ValueError: In debug mode, if name appears multiple times in trace
       """
       if not self.active:
           dist.sample()
           return dist.val

       if __debug__:
           if name in self.output_trace:
               raise ValueError(f"Name '{name}' in trace is not unique")

       if name in self.input_trace:
           trace_dist = self.input_trace[name]
           if ((trace_dist.fun.__name__, trace_dist.args, trace_dist.kwargs) == 
               (dist.fun.__name__, dist.args, dist.kwargs)):
               dist.val = trace_dist.val
           else:
               dist.repair(trace_dist)
       else:
           dist.sample()

       self.output_trace[name] = dist
       return dist.val

   def play(self, gen, trace):
       """
       Run a generator function with tracing.
       
       Activates tracing, runs the generator using the provided trace as input,
       collects the output trace, and returns both the solution and new trace.
       
       Args:
           gen (callable): Generator function to run
           trace (dict): Input trace to use for replay
           
       Returns:
           tuple: (solution from generator, output trace dictionary)
       """
       self.active = True
       self.input_trace = trace
       self.output_trace = {}

       solution = gen()
       
       result_trace = self.output_trace
       self.active = False
       return solution, result_trace

tracer = Tracer()

File: pto-webapp/src/tracer.js
/**
 * tracer.js — Trace recording and replay engine.
 *
 * The tracer is the heart of PTO. It sits between the generator function and
 * the random primitives (rnd), intercepting every random decision and recording
 * it in a "trace" — a dictionary mapping location keys to Dist objects.
 *
 * How it works:
 *   1. When the tracer is INACTIVE, rnd calls just sample normally.
 *   2. When the tracer is ACTIVE (during play()), each rnd call goes through
 *      tracer.sample(name, dist), which either:
 *        a) Reuses a value from a previous trace (if the name exists and the
 *           distribution parameters match),
 *        b) Repairs the value (if the name exists but parameters changed), or
 *        c) Samples a fresh value (if the name is new).
 *
 * Linear naming: When no explicit name is given, the tracer assigns sequential
 * integer keys (0, 1, 2, ...). This is the simplest scheme and works for any
 * generator. Structural naming (Slice 2) will inject explicit name strings
 * via AST transformation.
 *
 * The trace (genotype) fully determines the solution (phenotype) when replayed
 * through the same generator. This is what makes PTO's operators work: mutate
 * or crossover the trace, replay it, and get a new valid solution.
 */

export class Tracer {
  constructor() {
    /** Whether trace recording is currently active. */
    this.active = false;

    /** Trace from a previous solution, used to guide replay. */
    this.inputTrace = {};

    /** Trace being built during the current play() call. */
    this.outputTrace = {};

    /** Counter for linear (auto-increment) naming. */
    this.counter = 0;
  }

  /**
   * Record or replay a single random decision.
   *
   * Called by rnd methods. When active, this either reuses a value from the
   * input trace (if the name matches and distribution parameters agree),
   * repairs the value (if parameters changed), or samples fresh.
   *
   * @param {string|number|null} name  Trace key. null → auto-assign integer.
   * @param {Dist} dist  Distribution object (not yet sampled).
   * @returns {*}  The sampled/replayed value.
   */
  sample(name, dist) {
    if (!this.active) {
      dist.sample();
      return dist.val;
    }

    // Auto-assign linear name if none provided
    if (name == null) {
      name = this.counter++;
    }

    // Look up in input trace
    if (name in this.inputTrace) {
      const traceDist = this.inputTrace[name];
      if (dist.matches(traceDist)) {
        // Same distribution parameters → reuse cached value
        dist.val = traceDist.val;
      } else {
        // Parameters changed → repair (coarse: re-sample)
        dist.repair(traceDist);
      }
    } else {
      // New entry → sample fresh
      dist.sample();
    }

    this.outputTrace[name] = dist;
    return dist.val;
  }

  /**
   * Run a generator function under tracing, producing a solution and its trace.
   *
   * @param {Function} generator  Generator function that calls rnd methods.
   * @param {Object}   trace      Input trace to replay from (empty {} for fresh).
   * @returns {{ pheno: *, geno: Object }}  The phenotype and its genotype trace.
   */
  play(generator, trace) {
    this.active = true;
    this.inputTrace = trace;
    this.outputTrace = {};
    this.counter = 0;

    let pheno;
    try {
      pheno = generator();
    } finally {
      this.active = false;
    }

    return { pheno, geno: this.outputTrace };
  }
}
